File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/mailbox/mail_day/MailDayViewModel.kt
@file:Suppress("PackageNaming", "TooManyFunctions")

package app.pantopus.android.ui.screens.mailbox.mail_day

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.pantopus.android.data.api.models.mailbox.v2.PendingItemDto
import app.pantopus.android.data.api.models.mailbox.v2.ResolveRoutingRequest
import app.pantopus.android.data.api.net.NetworkResult
import app.pantopus.android.data.mailbox.MailboxRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.util.Locale
import javax.inject.Inject

/** Nav-arg key for the variant (`populated` or `empty`) — kept for the
 *  deep-link route; the live screen now derives the frame from the data. */
const val MAIL_DAY_VARIANT_KEY = "variant"

private const val UNDO_SECONDS = 5
private const val CONFIDENCE_SCALE = 100

/**
 * A13.16 — Backs the My Mail Day triage editor.
 *
 * `load()` fetches the unresolved routing queue from
 * `GET /api/mailbox/v2/pending` and renders each item as an "needs a call"
 * card. Accepting a suggestion optimistically moves the card into the
 * reviewed list and calls `POST /api/mailbox/v2/resolve`, rolling back on
 * failure. Finishing the day logs `POST /api/mailbox/v2/event` telemetry.
 *
 * Counts (routed / junked / returned / remaining / total) are derived
 * client-side — no summary endpoint returns them. Junk / return-to-sender
 * and per-item undo have no backend route, so those affordances remain
 * out of scope (the reviewed list only grows via accepted suggestions).
 */
@HiltViewModel
class MailDayViewModel
    @Inject
    constructor(
        private val repository: MailboxRepository,
    ) : ViewModel() {
        private val _state = MutableStateFlow<MailDayUiState>(MailDayUiState.Loading)
        val state: StateFlow<MailDayUiState> = _state.asStateFlow()

        private var onScanRequested: () -> Unit = {}

        fun configure(onScanRequested: () -> Unit) {
            this.onScanRequested = onScanRequested
        }

        fun load() {
            _state.value = MailDayUiState.Loading
            viewModelScope.launch {
                _state.value =
                    when (val result = repository.pending()) {
                        is NetworkResult.Success -> project(result.data.pending)
                        is NetworkResult.Failure -> MailDayUiState.Error(result.error.message)
                    }
            }
        }

        fun refresh() = load()

        /**
         * Build a render frame from the pending queue. The reviewed list
         * starts empty (it grows as the user triages this session); the
         * yesterday-recap + setup-nudges + streak surfaces have no source
         * on `/pending`, so they are left empty and the screen hides them.
         */
        private fun project(pending: List<PendingItemDto>): MailDayUiState {
            val unreviewed = pending.map(::mapItem)
            val content =
                MailDayContent(
                    dateLabel = todayLabel(),
                    streakDays = 0,
                    lastScanLabel = "",
                    unreviewed = unreviewed,
                    reviewed = emptyList(),
                    yesterdayRecap = null,
                    setupNudges = emptyList(),
                )
            return if (unreviewed.isEmpty()) MailDayUiState.Empty(content) else MailDayUiState.Populated(content)
        }

        private fun mapItem(dto: PendingItemDto): UnreviewedMailDayItem {
            val mail = dto.mail
            return UnreviewedMailDayItem(
                id = dto.mailId,
                kind = kindFor(mail?.mailObjectType, mail?.category),
                label = mail?.subject?.takeIf { it.isNotBlank() } ?: "Mail",
                sender = mail?.senderDisplay ?: mail?.senderBusinessName ?: "Unknown sender",
                suggestedName = dto.recipientNameRaw.orEmpty(),
                suggestedAvatar = MailDaySuggestedAvatar.PersonalSky,
                confidencePercent = ((dto.bestMatchConfidence ?: 0.0) * CONFIDENCE_SCALE).toInt().coerceIn(0, CONFIDENCE_SCALE),
                secondaryLabel = "Other",
            )
        }

        private fun kindFor(
            objectType: String?,
            category: String?,
        ): MailDayKind {
            if (category?.contains("bill", ignoreCase = true) == true) return MailDayKind.Bill
            return when (objectType) {
                "package" -> MailDayKind.Package
                "postcard" -> MailDayKind.Postcard
                "booklet" -> MailDayKind.Magazine
                "bundle" -> MailDayKind.Flyer
                else -> MailDayKind.Envelope
            }
        }

        private fun todayLabel(): String = LocalDate.now().format(DateTimeFormatter.ofPattern("EEEE, MMM d", Locale.US))

        fun requestScan() {
            onScanRequested()
        }

        /**
         * Accept the AI suggestion: optimistically move the card to the
         * reviewed list, then resolve the routing on the backend. On failure
         * the previous content is restored.
         */
        fun acceptSuggestion(id: String) {
            val current = _state.value as? MailDayUiState.Populated ?: return
            val target = current.content.unreviewed.firstOrNull { it.id == id } ?: return
            val previous = current.content
            val firstName = target.suggestedName.substringBefore(" ", target.suggestedName)
            val drawer =
                when (target.suggestedAvatar) {
                    MailDaySuggestedAvatar.PersonalSky -> "personal"
                    MailDaySuggestedAvatar.HouseholdGreen -> "home"
                }
            val newReviewed =
                listOf(
                    ReviewedMailDayItem(
                        id = target.id,
                        kind = target.kind,
                        label = target.label,
                        action = ReviewedMailAction.Routed,
                        routedTo = firstName,
                        routedTint =
                            when (target.suggestedAvatar) {
                                MailDaySuggestedAvatar.PersonalSky -> MailDayRoutedTint.PersonPrimary
                                MailDaySuggestedAvatar.HouseholdGreen -> MailDayRoutedTint.HouseholdHome
                            },
                        whenLabel = "just now",
                        undoCountdown = UNDO_SECONDS,
                    ),
                ) + current.content.reviewed.map { it.copy(undoCountdown = null) }
            _state.value =
                MailDayUiState.Populated(
                    current.content.copy(
                        unreviewed = current.content.unreviewed.filterNot { it.id == id },
                        reviewed = newReviewed,
                    ),
                )
            viewModelScope.launch {
                val result = repository.resolve(ResolveRoutingRequest(mailId = id, drawer = drawer))
                if (result is NetworkResult.Failure) {
                    _state.value = MailDayUiState.Populated(previous)
                }
            }
        }

        /** Finish-day commit — no persistence route exists, so log telemetry. */
        fun finishDay() {
            if (!canFinishDay) return
            viewModelScope.launch {
                repository.logEvent(
                    eventType = "mailday_finished",
                    metadata =
                        mapOf(
                            "routed" to routedCount.toString(),
                            "junked" to junkedCount.toString(),
                            "returned" to returnedCount.toString(),
                        ),
                )
            }
        }

        fun tickUndo() {
            val current = _state.value as? MailDayUiState.Populated ?: return
            val reviewed = current.content.reviewed.toMutableList()
            val index = reviewed.indexOfFirst { (it.undoCountdown ?: 0) > 0 }
            if (index == -1) return
            val next = (reviewed[index].undoCountdown ?: 0) - 1
            reviewed[index] = reviewed[index].copy(undoCountdown = if (next > 0) next else null)
            _state.value = MailDayUiState.Populated(current.content.copy(reviewed = reviewed))
        }

        // ---- Derived counters used by the FinishDay bar + DayHeader ----

        val total: Int
            get() {
                val populated = _state.value as? MailDayUiState.Populated ?: return 0
                return populated.content.unreviewed.size + populated.content.reviewed.size
            }

        val done: Int
            get() = (_state.value as? MailDayUiState.Populated)?.content?.reviewed?.size ?: 0

        val remaining: Int
            get() = (_state.value as? MailDayUiState.Populated)?.content?.unreviewed?.size ?: 0

        val routedCount: Int
            get() =
                (_state.value as? MailDayUiState.Populated)
                    ?.content
                    ?.reviewed
                    ?.count { it.action == ReviewedMailAction.Routed } ?: 0

        val junkedCount: Int
            get() =
                (_state.value as? MailDayUiState.Populated)
                    ?.content
                    ?.reviewed
                    ?.count { it.action == ReviewedMailAction.Junked } ?: 0

        val returnedCount: Int
            get() =
                (_state.value as? MailDayUiState.Populated)
                    ?.content
                    ?.reviewed
                    ?.count { it.action == ReviewedMailAction.Returned } ?: 0

        val canFinishDay: Boolean
            get() = _state.value is MailDayUiState.Populated && remaining == 0 && total > 0
    }

File: frontend/apps/ios/Pantopus/Features/Mailbox/MailDay/MailDayViewModel.swift
//
//  MailDayViewModel.swift
//  Pantopus
//
//  A13.16 — Backs the My Mail Day editor. Two variants seeded by the
//  caller:
//
//    `.populated` — mid-afternoon triage. The latest reviewed row pulses
//      a 5-second undo countdown; `tickUndo()` (driven by `MailDayView`)
//      decrements the seconds and clears the chip when it hits 0.
//    `.empty` — no scans yet today; the hero + recap + setup-nudges
//      render off the same `MailDayContent`.
//
//  Backend has been removed from the repo, so `load()` projects a
//  deterministic fixture (`MailDaySampleData`). Action handlers
//  (`route` / `junk` / `undo`) flip in-memory state — they don't write
//  to the wire — and exist so the screen feels real in previews.
//

import Foundation
import Observation

public enum MailDayVariant: String, Sendable, Hashable {
    case populated
    case empty
}

@Observable
@MainActor
public final class MailDayViewModel {
    public private(set) var state: MailDayState = .loading

    public var variant: MailDayVariant
    private let seededContent: MailDayContent?
    private let onScanRequested: @MainActor () -> Void

    /// - Parameters:
    ///   - variant: Which fixture to project. Defaults to `.populated`.
    ///   - content: Optional seed (tests / previews) overriding the
    ///     sample fixture for this variant.
    ///   - onScanRequested: Invoked when the user taps any Scan CTA
    ///     (top scan-more card or empty-hero primary). Out of scope to
    ///     wire to the real scanner here — the host hands a closure.
    public init(
        variant: MailDayVariant = .populated,
        content: MailDayContent? = nil,
        onScanRequested: @escaping @MainActor () -> Void = {}
    ) {
        self.variant = variant
        seededContent = content
        self.onScanRequested = onScanRequested
    }

    // MARK: - Lifecycle

    public func load() async {
        state = projectedState()
    }

    public func refresh() async {
        state = projectedState()
    }

    private func projectedState() -> MailDayState {
        switch variant {
        case .populated:
            .populated(seededContent ?? MailDaySampleData.populated)
        case .empty:
            .empty(seededContent ?? MailDaySampleData.empty)
        }
    }

    // MARK: - Derived

    /// Total pieces (reviewed + unreviewed) — drives the ProgressRing
    /// denominator + the "Finish day · N pieces" CTA label.
    public var total: Int {
        guard case let .populated(content) = state else { return 0 }
        return content.unreviewed.count + content.reviewed.count
    }

    /// Pieces with a decision (any action) — ProgressRing numerator.
    public var done: Int {
        guard case let .populated(content) = state else { return 0 }
        return content.reviewed.count
    }

    /// Pieces still needing a call. Drives the "Finish day" disabled
    /// state.
    public var remaining: Int {
        guard case let .populated(content) = state else { return 0 }
        return content.unreviewed.count
    }

    /// Routed-only subset for the FinishDay summary line.
    public var routedCount: Int {
        guard case let .populated(content) = state else { return 0 }
        return content.reviewed.filter { $0.action == .routed }.count
    }

    public var junkedCount: Int {
        guard case let .populated(content) = state else { return 0 }
        return content.reviewed.filter { $0.action == .junked }.count
    }

    public var returnedCount: Int {
        guard case let .populated(content) = state else { return 0 }
        return content.reviewed.filter { $0.action == .returned }.count
    }

    /// `true` once every piece has a decision and FinishDay can fire.
    public var canFinishDay: Bool {
        guard case .populated = state else { return false }
        return remaining == 0 && total > 0
    }

    // MARK: - Actions

    /// Out-of-scope per the task — open the scanner flow.
    public func requestScan() {
        onScanRequested()
    }

    /// Decrement the undo countdown on the latest reviewed row. When it
    /// hits 0 the chip clears and the row reads as a normal reviewed
    /// entry. Idempotent on zero / nil.
    public func tickUndo() {
        guard case let .populated(content) = state else { return }
        var updated = content.reviewed
        guard let firstIndex = updated.firstIndex(where: { ($0.undoCountdown ?? 0) > 0 }) else {
            return
        }
        let current = updated[firstIndex]
        let next = (current.undoCountdown ?? 0) - 1
        updated[firstIndex] = ReviewedMailDayItem(
            id: current.id,
            kind: current.kind,
            label: current.label,
            action: current.action,
            routedTo: current.routedTo,
            routedTint: current.routedTint,
            whenLabel: current.whenLabel,
            undoCountdown: next > 0 ? next : nil
        )
        state = .populated(
            MailDayContent(
                dateLabel: content.dateLabel,
                streakDays: content.streakDays,
                lastScanLabel: content.lastScanLabel,
                unreviewed: content.unreviewed,
                reviewed: updated,
                yesterdayRecap: content.yesterdayRecap,
                setupNudges: content.setupNudges
            )
        )
    }

    /// Move an unreviewed piece into the reviewed list as `.routed` —
    /// the AI suggestion is treated as accepted.
    public func acceptSuggestion(for itemId: String) {
        guard case let .populated(content) = state else { return }
        guard let index = content.unreviewed.firstIndex(where: { $0.id == itemId }) else { return }
        let item = content.unreviewed[index]
        var unreviewed = content.unreviewed
        unreviewed.remove(at: index)
        let firstName = item.suggestedName.split(separator: " ").first.map(String.init) ?? item.suggestedName
        let reviewed = [
            ReviewedMailDayItem(
                id: item.id,
                kind: item.kind,
                label: item.label,
                action: .routed,
                routedTo: firstName,
                routedTint: item.suggestedAvatar == .personalSky ? .personPrimary : .householdHome,
                whenLabel: "just now",
                undoCountdown: 5
            )
        ] + clearedReviewed(content.reviewed)
        state = .populated(
            MailDayContent(
                dateLabel: content.dateLabel,
                streakDays: content.streakDays,
                lastScanLabel: content.lastScanLabel,
                unreviewed: unreviewed,
                reviewed: reviewed,
                yesterdayRecap: content.yesterdayRecap,
                setupNudges: content.setupNudges
            )
        )
    }

    /// Clear `undoCountdown` from every prior row so only the latest
    /// action shows the pulse — matches the JSX's single-row treatment.
    private func clearedReviewed(_ rows: [ReviewedMailDayItem]) -> [ReviewedMailDayItem] {
        rows.map { row in
            ReviewedMailDayItem(
                id: row.id,
                kind: row.kind,
                label: row.label,
                action: row.action,
                routedTo: row.routedTo,
                routedTint: row.routedTint,
                whenLabel: row.whenLabel,
                undoCountdown: nil
            )
        }
    }
}
